internal/gcsx/file_cache_reader_test.go

```diff
@@ -81,3 +98,209 @@ func (t *FileCacheReaderTest) TestNewFileCacheReader() {
+	start2 := 16*util.MiB + 4
+	end2 := start2 + util.MiB
+	buf2 := make([]byte, end2-start2)
+
+	readerResponse, err = t.reader.ReadAt(t.ctx, buf2, int64(start2))
+
+	// Assuming a download with a start offset of start2 is in progress, a fallback to another reader will be required.
+	assert.True(t.T(), errors.Is(err, FallbackToAnotherReader))
+	assert.Zero(t.T(), readerResponse.Size)
+	job := t.jobManager.GetJob(t.object.Name, t.mockBucket.Name())
+	assert.True(t.T(), job == nil || job.GetStatus().Name == downloader.Downloading)
+	t.mockBucket.AssertExpectations(t.T())
+}
+
+func (t *fileCacheReaderTest) Test_ReadAt_SequentialToRandomSubsequentReadOffsetLessThanPrevious() {
+	t.object.Size = 20 * util.MiB
+	testContent := testutil.GenerateRandomBytes(int(t.object.Size))
+	rd := &fake.FakeReader{ReadCloser: getReadCloser(testContent)}
+	t.mockNewReaderWithHandleCallForTestBucket(0, t.object.Size, rd)
+	t.mockBucket.On("Name").Return("test-bucket")
+	start1 := 0
+	end1 := util.MiB
+	assert.Less(t.T(), end1, int(t.object.Size))
+	// First call from offset 0 - sequential read
+	buf := make([]byte, end1-start1)
+	readerResponse, err := t.reader.ReadAt(t.ctx, buf, int64(start1))
+	// Served from file cache
+	assert.NoError(t.T(), err)
+	assert.Equal(t.T(), readerResponse.DataBuf, testContent[start1:end1])
+	start2 := 16*util.MiB + 4
+	end2 := start2 + util.MiB
+	buf2 := make([]byte, end2-start2)
+	// Assuming a download with a start offset of start2 is in progress, a fallback to another reader will be required.
+	readerResponse, err = t.reader.ReadAt(t.ctx, buf2, int64(start2))
+	assert.True(t.T(), errors.Is(err, FallbackToAnotherReader))
+	assert.Zero(t.T(), readerResponse.Size)
+	// Assuming start3 offset is downloaded
+	start3 := 4 * util.MiB
+	end3 := start3 + util.MiB
+	buf3 := make([]byte, end3-start3)
+
+	readerResponse, err = t.reader.ReadAt(t.ctx, buf3, int64(start3))
+
+	assert.NoError(t.T(), err)
+	assert.Equal(t.T(), readerResponse.DataBuf, testContent[start3:end3])
+	t.mockBucket.AssertExpectations(t.T())
+}
+
+func (t *fileCacheReaderTest) Test_ReadAt_CacheMissDueToInvalidJob() {
+	testContent := testutil.GenerateRandomBytes(int(t.object.Size))
+	rc1 := &fake.FakeReader{ReadCloser: getReadCloser(testContent)}
+	t.mockNewReaderWithHandleCallForTestBucket(0, t.object.Size, rc1)
+	t.mockBucket.On("Name").Return("test-bucket")
+	buf := make([]byte, t.object.Size)
+	readerResponse, err := t.reader.ReadAt(t.ctx, buf, 0)
+	assert.NoError(t.T(), err)
+	assert.Equal(t.T(), readerResponse.DataBuf, testContent)
+	job := t.jobManager.GetJob(t.object.Name, t.mockBucket.Name())
+	if job != nil {
+		jobStatus := job.GetStatus().Name
+		assert.True(t.T(), jobStatus == downloader.Downloading || jobStatus == downloader.Completed, fmt.Sprintf("the actual status is %v", jobStatus))
+	}
+
+	err = t.reader.fileCacheHandler.InvalidateCache(t.object.Name, t.mockBucket.Name())
+	assert.NoError(t.T(), err)
+
+	readerResponse, err = t.reader.ReadAt(t.ctx, buf, 0)
```

Comment:
Lets follow the AAA - assuming we are testing the ReadAt() method, but seems we are testing InvalidateCache method.